### ISSUE TITLE
Dynamically alter CSP

### ISSUE THREAD

**Issue by demianbrecht** [pypa/warehouse#1005]
Currently CSP is set statically in warehouse/config.py. We'll need to add a mechanism for views or templates to alter this, preferably in some generalized fashion.

Use case: I'm adding recaptcha to the user registration page. Amongst source additions to `script-src` et al, it also requires the addition of `"unsafe-inline"` to `style-src` (see [here](https://developers.google.com/recaptcha/docs/faq#recently-my-recaptcha-widget-started-displaying-invalid-site-key-whats-happening for details)). Obviously, this isn't something that we want to add statically for every page and should only be added when absolutely necessary.

A couple different methods I can think of to potentially solve this:

- Add nonces to the *-src headers and make them accessible to templates. This would mean moving the CSP policy formatting to the tween function rather than being in the factory method which would cause a per-render performance impact (albeit super minor). This also wouldn't solve the issue of "unsafe-inline" or other header values not directly rendered in templates.
- Add a decorator to add header value changes as needed per view. This will require moving from setting `request.response.headers` in warehouse.config explicitly to using `request.response.headerlist` so that multiple `Content-Security-Policy` headers can be included in the response. Based on what I've read, browsers *should* support intersecting multiple headers, but that'll have to actually be tested.

Thoughts?

(Maybe there's something already allowing for all this in warehouse and I just haven't found it, which would be wonderful ;))

**Comment by di** [comment]
(I think you actually meant to link to https://developers.google.com/recaptcha/docs/faq#im-using-content-security-policy-csp-on-my-website-how-can-i-configure-it-to-work-with-recaptcha)

I don't think nonces would work here, because the styles/scripts are coming from Google, not from somewhere where we would be able to set a nonce on them.

I don't really see a better way to solve this than to allow `style-src 'unsafe-inline'` on any page using the recaptcha, which probably means any recaptcha should be on it's own page.

**Comment by dstufft** [comment]
Maybe what we want is to move to ``warehouse/csp.py`` and do something like:

```python
def content_security_policy_tween_factory(handler, registry):
    def content_security_policy_tween(request):
        resp = handler(request)

        # We don't want to apply our Content Security Policy to the debug
        # toolbar, that's not part of our application and it doesn't work with
        # our restrictive CSP.
        if request.path.startswith("/_debug_toolbar/"):
            return resp

        policy = registry.settings.get("csp", {})
        policy = some_hypothetical_dict_merge(
            policy,
            request.find_service(name="csp"),
        )

        policy = "; ".join([
            " ".join([k] + [v2 for v2 in v if v2 is not None])
            for k, v in sorted(policy.items())
            if [v2 for v2 in v if v2 is not None]
        ])

        resp.headers["Content-Security-Policy"] = policy.format(request=request)

        return resp

    return content_security_policy_tween


def csp_factory(request, context):
    return {}


def includeme(config):
    config.add_service_factory(csp_factory, name="csp")
    config.add_tween("warehouse.csp.content_security_policy_tween_factory")
```

Then have view code do like:

```python
def my_view(request):
    request.find_service(name="csp")["style-src"] = ["unsafe-inline"]
    ...
```

That's all completely untested and doesn't actually work, but would maybe give an idea of what could work out pretty OK? I think that ``request.find_service(name="csp")`` will cache the dictionary it creates on the first access and then return the same one each time, but if it doesn't (or if you don't like that!), you could probably also use ``config.add_request_method(lambda request: {}, name="csp", reify=True)`` and then just use ``request.csp`` as the dictionary to store additional stuff in.

It might even make sense to remove the merge thing, and just make the default value a deep copy of the csp settings dictionary. That would depend on whether we'd want a view to be able to *remove* default settings, or if it should only be additive. Off the top of my head, I think it should only be additive because I can't think of a use case where you'd want to say, remove the default CSS on one particular page.

**Comment by dstufft** [comment]
If it's not obvious from above, I'm not really worried about the rendering cost of running some extra code on every request here. It'll be completely undetectable for a human, and the vast bulk of our pages will be served out of the Fastly cache anyways.

**Comment by demianbrecht** [comment]
@dstufft yeah I think something like that would work and would be pretty clean.

@di unless I'm misunderstanding the spec, nonces would work for the script tag that's embedded in the template. But you're right that it won't solve all the problems here as `"unsafe-inline"` is required. I think what @dstufft has proposed will solve both cases. 

I'll work on putting a change together for this as it's a requirement for the first iteration of the registration form.

**Comment by di** [comment]
Great, we can use it to make the XML/RSS look not-so-awful as well :smile:

**Comment by dstufft** [comment]
Fixed via #1007.
